Refactor ISyntaxFactsService.IsQueryKeyword

MaStr11 opened the pull request:
### Customer scenario

This is a follow up to #23049. That PR introduced the `bool IsQueryKeyword(SyntaxToken token);` method in  `ISyntaxFactsService` in commit 2f160258186111147dbe0f24738bc8bfbda23cb8. The way the methods checks the keyword might break if any of the keywords is allowed in other contexts. This was revealed in https://github.com/dotnet/roslyn/pull/25711#discussion_r176950992 by @CyrusNajmabadi. This PR fixes this by checking if the parent SyntaxNode of each keyword token is somehow related to query syntax for `CSharpSyntaxFactsService` and `VisualBasicSyntaxFactsServiceFactory`.

~Unfortunately I can't provide tests with this change because there isn't any test infrastructure in place for `ISyntaxFactsService`. The way I *tested* the code was by looking for the usage of the keywords in the `SyntaxFactory` of VB and C#. That way I found the places where the keywords got used and their parents type.~

Update: VB and C# implementations are covered by tests now.

### Bugs this fixes

None.

### Workarounds, if any

None.

### Risk

Low. This method is only called by [GetSemanticInfo.SemanticModelExtensions](https://github.com/dotnet/roslyn/blob/master/src/Workspaces/Core/Portable/Shared/Extensions/SemanticModelExtensions.cs#L238).

### Performance impact

Low.

### Is this a regression from a previous update?

Just a re-factoring.

### Root cause analysis

Found via code review after PR was already merged.

### How was the bug found?

Found by @CyrusNajmabadi 

/cc @Neme12

CyrusNajmabadi commented on src/Workspaces/VisualBasic/Portable/LanguageServices/VisualBasicSyntaxFactsService.vb:
```diff
@@ -303,53 +303,32 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
-                    Return False
+                    Return TypeOf token.Parent Is QueryClauseSyntax
+                Case SyntaxKind.GroupKeyword
+                    Return (TypeOf token.Parent Is QueryClauseSyntax) OrElse (token.Parent.IsKind(SyntaxKind.GroupAggregation))
```
odd

MaStr11 replied:
This refers to

```VB
        Public Shared Function GroupAggregation() As GroupAggregationSyntax
            Return SyntaxFactory.GroupAggregation(SyntaxFactory.Token(SyntaxKind.GroupKeyword))
        End Function
```

and 

```VB
        Public Shared Function GroupByClause(items As SeparatedSyntaxList(Of ExpressionRangeVariableSyntax), keys As SeparatedSyntaxList(Of ExpressionRangeVariableSyntax), aggregationVariables As SeparatedSyntaxList(Of AggregationRangeVariableSyntax)) As GroupByClauseSyntax
            Return SyntaxFactory.GroupByClause(SyntaxFactory.Token(SyntaxKind.GroupKeyword), items, SyntaxFactory.Token(SyntaxKind.ByKeyword), keys, SyntaxFactory.Token(SyntaxKind.IntoKeyword), aggregationVariables)
        End Function

        Public Shared Function GroupJoinClause(joinedVariables As SeparatedSyntaxList(Of CollectionRangeVariableSyntax), additionalJoins As SyntaxList(of JoinClauseSyntax), joinConditions As SeparatedSyntaxList(Of JoinConditionSyntax), aggregationVariables As SeparatedSyntaxList(Of AggregationRangeVariableSyntax)) As GroupJoinClauseSyntax
            Return SyntaxFactory.GroupJoinClause(SyntaxFactory.Token(SyntaxKind.GroupKeyword), SyntaxFactory.Token(SyntaxKind.JoinKeyword), joinedVariables, additionalJoins, SyntaxFactory.Token(SyntaxKind.OnKeyword), joinConditions, SyntaxFactory.Token(SyntaxKind.IntoKeyword), aggregationVariables)
        End Function
```

Neme12 replied:
It's a little odd that you're mixing kind and type checks here.

MaStr11 commented:
Ping. This PR needs a buddy.

jasonmalinowski commented on src/Workspaces/VisualBasic/Portable/LanguageServices/VisualBasicSyntaxFactsService.vb:
```diff
@@ -303,53 +303,32 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
-                        Case Else
-                            Return False ' e.g. ForEach ... in
-                    End Select
+                Case SyntaxKind.JoinKeyword
```
Don't these need to be listed comma-delimited? Note that C# and VB syntaxes here are different -- it's not a "fall through" model like C#.

jasonmalinowski commented on src/Workspaces/VisualBasic/Portable/LanguageServices/VisualBasicSyntaxFactsService.vb:
```diff
@@ -303,53 +303,32 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+                Case SyntaxKind.InKeyword
+                    Return TypeOf token.Parent Is CollectionRangeVariableSyntax
             End Select
+            Return False
```
Put this inside "Case Else" to give some automated tools a chance to catch your fall-through bug.

MaStr11 commented:
> Given this nearly introduced a bug, let's add tests.

from the description in the PR:

> Unfortunately I can't provide tests with this change because there isn't any test infrastructure in place for `ISyntaxFactsService`. The way I *tested* the code was by looking for the usage of the keywords in the SyntaxFactory of VB and C#. That way I found the places where the keywords got used and their parents type.

Can you tell me how/where the test should go?

MaStr11 commented:
@jasonmalinowski I fixed the Select Case fall-through bug and added some tests for VB. I added tests for all keywords (each keyword is at least tested once, but not every possible position is tested). I also added a few tests for the `while` keyword being in another context (loops). I could add more of those tests for other keywords (e.g. `On` being used in `On Error Goto`) if you want me to do so. I also haven't added any C# tests (It seems there isn't any `CSharpSyntaxFactsServiceTest` class).

jasonmalinowski commented:
@dotnet-bot retest this please.

jasonmalinowski commented:
@jinujoseph Tagging for delegated M2 approval. Just a preventative bug fix @CyrusNajmabadi found.

jasonmalinowski commented:
@dotnet-bot retest this please.

jasonmalinowski commented:
@jinujoseph Test failures appear to be unrelated; we'll drive them but at this point this would be waiting for your approval.

jinujoseph commented:
Approved to merge for 15.8.Preview3

CyrusNajmabadi commented:
> Unfortunately I can't provide tests with this change because there isn't any test infrastructure in place for ISyntaxFactsService

Note: 
1. i don't think that's true. :)
2. even if it's true that we don't have *direct* unit tests of this interface, we can certain test it through wahtever features eventually call *through* to this service right?  After all, this method must exist for some feature to use.  

In general, i prefer the latter BTW.  These types of services are internal impl details.  Tey can change their shape and impl, but would not necessarily require any tests or features to change behavior.

Neme12 commented:
I agree, this should have been caught by tests for the features that are using this. They can change their impl any time to not use this helper anymore and can easily break that way.

MaStr11 commented:
The problem is that this particular method was used to suppress Quickinfo for the `select` keyword in cases where the `select` isn't mapped to a call to `Select` ([degenerated expression](https://docs.microsoft.com/de-de/dotnet/csharp/language-reference/language-specification/expressions#degenerate-query-expressions)). So the method could also be named `IsSelectKeyword` because it is never called in other circumstances.

jasonmalinowski commented:
Just waiting on some test issues to be fixed before merging.

Neme12 commented:
> it is never called in other circumstances

I don't understand. So what's the purpose of this method and all the logic if all we need is `IsSelectKeyword`?

MaStr11 commented:
The method is called [here](https://github.com/dotnet/roslyn/blob/5d5d9d5295f29712eff03dda9665ab095fdd1446/src/Workspaces/Core/Portable/Shared/Extensions/SemanticModelExtensions.cs#L237). At that point we are asking *if you can't find a symbol at the current cursor position and you are at a query keyword, then don't show any quick info*. This behavior was part of the feature specification #23394.
At the time of writing it wasn't totally clear to me which keywords are not mapped to symbols. It turned out that only three keywords are not mapped:

* degenerated 'select'
* first 'from' if not followed by a type
* query continuation with 'into' after a 'select' (I'm not sure about this one)

So the feature that uses the method can never test every branch of IsQueryKeyword because most keywords are mapped and 'allSymbols.Length' is <>0.

I just followed the feature specification that said:

> For those clauses that don't directly map to a method a hint is shown: "This clause isn't mapped to an method call due to optimizations." or if this is not possible no quick info is shown.

I hope that gives some context and answers your question.

jasonmalinowski commented:
@MaStr11 Are we better off then just simplifying this to the methods we actually need? I don't have any strong opinions here, just making sure we all think this is the right thing. Unless @CyrusNajmabadi or @Neme12 have any other thoughts I'll pass this back off for final approval.

CyrusNajmabadi commented:
i had to be a negative-nancy.  But if this really is just "IsSelectKeyword", i thnik i'd much rather have this just become that method, and not be in ISyntaxFacts at all.  Just have the specialized logic for whatever part of quickinfo needs it.

MaStr11 commented:
From the feature specification I can tell that the method gets called in the following circumstances:

* degenerated 'select'
* first 'from' if not followed by a type
* query continuation with 'into' after a 'select' (I'm not sure about this one)
* 'orderby', if the first argument is followed by 'ascending' or 'descending'.

There might be even more cases.

This method is needed to suppress quickinfo for these keywords. If we wouldn't do this quick info would show the resulting type of the whole linq expression, which would be an odd behavior.

I think it would be problematic to find a good name for the method given if we would only special case the mentioned occurrences. Furthermore the method fits  the feature specification that just says: if you can't find a method that the keyword maps to and the keyword belongs to a linq query then don't show quick info.

MaStr11 commented:
@CyrusNajmabadi I would thertenly be willing to add some C# tests as I did for VB. What do you think about that?

jasonmalinowski commented:
@CyrusNajmabadi, thoughts here?

MaStr11 commented:
I added the corresponding tests for C# now. Given that there are more Linq keywords than just `select` that need to be covered (see https://github.com/dotnet/roslyn/pull/25925#issuecomment-396328465) I think this is the best way forward.

MaStr11 commented:
@CyrusNajmabadi and @jasonmalinowski Any feedback would be appreciated. Thanks.

CyrusNajmabadi commented:
@MaStr11 i can look tonight/tomorrow!

MaStr11 commented:
Any updates?

MaStr11 commented:
@CyrusNajmabadi could you have a look please. The *IsQueryKeyword* function is fully covered by tests for C# and VB now. So now we make sure that the query keywords are really in a query context (to avoid regressions if a query keyword is used in other contexts in the future) and there are tests in place (because the only consuming feature of this method can only test a subset of the query keywords).

MaStr11 commented:
@jasonmalinowski Can you give an update please? (@CyrusNajmabadi isn't available the next few days btw.)

CyrusNajmabadi commented:
@jinujoseph @jasonmalinowski ^

jasonmalinowski commented:
@CyrusNajmabadi: hmm? did you have any further thoughts for this or is it good? @MaStr11 changed a few things and I wasn't sure if your signoff predated that.

CyrusNajmabadi commented:
Nope.  Still LGTM.  Thanks!
